internal/auth/service_list_auth_methods.go

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package auth
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/pagination"
+)
+
+// ListAuthMethods lists up to page size auth methods, filtering out entries that
+// do not pass the filter item function. It will automatically request
+// more auth methods from the database, at page size chunks, to fill the page.
+// It returns a new list token used to continue pagination or refresh items.
+// Auth methods are ordered by create time descending (most recently created first).
+func ListAuthMethods(
+	ctx context.Context,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn pagination.ListFilterFunc[AuthMethod],
+	repo *AuthMethodRepository,
+	scopeIds []string,
+) (*pagination.ListResponse[AuthMethod], error) {
+	const op = "auth.ListAuthMethods"
+
+	switch {
+	case len(grantsHash) == 0:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	case pageSize < 1:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "page size must be at least 1")
+	case filterItemFn == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing filter item callback")
+	case scopeIds == nil:
```

Comment:
Can we make this compare against the length of the slice instead? This will currently miss an empty, non-nil slice.